lib/llvmopencl/SafeDivision.cc

```diff
@@ -0,0 +1,184 @@
+                                       ConstantInt::get(Divisor->getType(), 0));
+  Value *FixedDivisor = Builder.CreateSelect(
+      Condition, ConstantInt::get(Divisor->getType(), 1), // true
+      Divisor);                                           // false
+
+#ifdef DEBUG_SAFE_DIVISION
+  std::cerr << "UNSIGNED fixing OP2: \n";
+  Fixed->dump();
+#endif
+
+  BinI->setOperand(1, FixedDivisor);
+  ZeroCheckedDivisorCache.insert(std::make_pair(Divisor, FixedDivisor));
+}
+
+static void fixSignedDivRem(IRBuilder<> &Builder, BinaryOperator *BinI,
+                            Value *Dividend, Value *Divisor,
+                            ValueToValueMapTy &ZeroCheckedCache,
+                            ValueToValueMapTy &IntMinCheckedCache) {
+  fixUnsignedDivRem(Builder, BinI, Dividend, Divisor, ZeroCheckedCache);
+
+  auto It = IntMinCheckedCache.find(Divisor);
+  if (It != IntMinCheckedCache.end()) {
+    BinI->setOperand(0, It->second);
+    return;
+  }
+
+  Type *T = Divisor->getType();
+  // for signed div/rem, we need to fix also the dividend (INT_MIN / -1)
+  Value *Condition1 = Builder.CreateCmp(CmpInst::Predicate::ICMP_EQ, Divisor,
+                                        ConstantInt::getSigned(T, -1));
+  // smallest negative value representable in given integer bit width
+  int64_t IntMin = (int64_t)1 << (T->getScalarSizeInBits() - 1);
+  Value *Condition2 = Builder.CreateCmp(CmpInst::Predicate::ICMP_EQ, Dividend,
+                                        ConstantInt::getSigned(T, IntMin));
+  Value *BothConds = Builder.CreateAnd(Condition1, Condition2);
+  Value *ConditionZExt = Builder.CreateZExt(BothConds, Dividend->getType());
+  Value *FixedDividend = Builder.CreateAdd(Dividend, ConditionZExt);
+
+#ifdef DEBUG_SAFE_DIVISION
+  std::cerr << "SIGNED fixing OP1: \n";
+  Fixed2->dump();
+#endif
+
+  BinI->setOperand(0, FixedDividend);
+  IntMinCheckedCache.insert(std::make_pair(Divisor, FixedDividend));
+}
+
+static bool handleDivisionUB(llvm::Function &F) {
+  IRBuilder<> Builder(F.getContext());
+
+#ifdef DEBUG_SAFE_DIVISION
+  std::cerr << "BEFORE:\n";
+  F.dump();
+#endif
+
+  // save already fixed values
+  ValueToValueMapTy ZeroCheckedCache;
+  ValueToValueMapTy IntMinCheckedCache;
+
+  bool Changed = false;
+  for (Function::iterator I = F.begin(), E = F.end(); I != E; ++I) {
+    for (BasicBlock::iterator BI = I->begin(), BE = I->end(); BI != BE;) {
+      BinaryOperator *BinI = dyn_cast<BinaryOperator>(BI++);
+      if (BinI == nullptr)
+        continue;
+      if (!BinI->isIntDivRem())
+        continue;
+      // TODO do we need to handle case of division by zero constant ?
+      if (dyn_cast<Constant>(BinI->getOperand(1)))
+        continue;
+#ifdef DEBUG_SAFE_DIVISION
+      std::cerr << "Fixing division/remainder in Inst: \n";
+      BinI->dump();
+#endif
+      Value *Dividend = BinI->getOperand(0);
+      Value *Divisor = BinI->getOperand(1);
+      // unlikely case but its possible optimization haven't run yet
```

Comment:
Yes, we have to be prepared to be ran on unoptimized Clang FE output.

Reply:
yep, will remove the comment